airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/xmin/XminCtidUtils.java

```diff
@@ -0,0 +1,111 @@
+package io.airbyte.integrations.source.postgres.xmin;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.collect.Sets;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.integrations.source.relationaldb.state.StateManager;
+import io.airbyte.protocol.models.v0.AirbyteStateMessage;
+import io.airbyte.protocol.models.v0.AirbyteStreamNameNamespacePair;
+import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog;
+import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream;
+import io.airbyte.protocol.models.v0.StreamDescriptor;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class XminCtidUtils {
+
+  public static StreamsCategorised categoriseStreams(final StateManager stateManager, final ConfiguredAirbyteCatalog fullCatalog) {
+    final List<AirbyteStateMessage> rawStateMessages = stateManager.getRawStateMessages();
+    final List<AirbyteStateMessage> statesFromCtidSync = new ArrayList<>();
+    final List<AirbyteStateMessage> statesFromXminSync = new ArrayList<>();
+
+    final Set<AirbyteStreamNameNamespacePair> alreadySeenStreams = new HashSet<>();
+    final Set<AirbyteStreamNameNamespacePair> streamsStillInCtidSync = new HashSet<>();
+
+    if (rawStateMessages != null) {
+      rawStateMessages.forEach(s -> {
+        final JsonNode streamState = s.getStream().getStreamState();
+        final StreamDescriptor streamDescriptor = s.getStream().getStreamDescriptor();
+        if (streamState == null || streamDescriptor == null) {
+          return;
+        }
+        final AirbyteStateMessage clonedState = Jsons.clone(s);
+        if (streamState.has("state_type") && streamState.get("state_type").asText().equalsIgnoreCase("ctid")) {
+          statesFromCtidSync.add(clonedState);
+          streamsStillInCtidSync.add(new AirbyteStreamNameNamespacePair(streamDescriptor.getName(), streamDescriptor.getNamespace()));
+        } else {
+          statesFromXminSync.add(clonedState);
+        }
+        alreadySeenStreams.add(new AirbyteStreamNameNamespacePair(streamDescriptor.getName(), streamDescriptor.getNamespace()));
+      });
+    }
+
+    final List<ConfiguredAirbyteStream> newlyAddedStreams = identifyNewlyAddedStreams(fullCatalog, alreadySeenStreams);
+    final List<ConfiguredAirbyteStream> streamsForCtidSync = new ArrayList<>();
+    fullCatalog.getStreams().stream()
+        .filter(stream -> streamsStillInCtidSync.contains(AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.getStream())))
+        .map(Jsons::clone)
+        .forEach(streamsForCtidSync::add);
+
+    streamsForCtidSync.addAll(newlyAddedStreams);
+
+    final List<ConfiguredAirbyteStream> streamsForXminSync = fullCatalog.getStreams().stream()
```

Comment:
I assume this is where the logic for the double wraparound will be added. 
If it's a double wraparound : 
- We need to add to the ctid streams instead of the xmin sync